### PR TITLE
Design refactor

### PR DESCRIPTION
**Pull request by nassiharel**
This change is [Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1469)

### CODE REVIEW

File: core/algorithm-queue/tests/test.js [hunk 1]
```diff
@@ -137,7 +137,7 @@ describe('Test', () => {
         });
     });
     describe('persistency tests', () => {
-        it('persistent load', async () => {
+        it.skip('persistent load', async () => {
```
Comment (golanha) [inline core/algorithm-queue/tests/test.js:140]: test left in skip

File: core/pipeline-driver-queue/tests/test.js [hunk 2]
```diff
@@ -174,12 +176,20 @@ describe('Test', () => {
...
+            const pipeline = {
+                name: 'test',
+                experimentName: 'test',
+            };
+            const status = {
+                status: 'queued'
+            };
+            const count = 50;
+            const keys = Array.from(Array(count).keys());
+            await Promise.all(keys.map(() => dataStore.createJob({ jobId, pipeline, status })));
```
Comment (golanha) [inline core/pipeline-driver-queue/tests/test.js:222]: Does the Datastore allow creating different records with the same jobId?

**Review by golanha — CHANGES_REQUESTED** [review]

**Review by golanha — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1469)* status: 0 of 82 files reviewed, 3 unresolved discussions (waiting on @NassiHarel and @yehiyam)

___
*[core/pipeline-driver-queue/lib/jobs/producer.js, line 67 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1469#-MraX1Dk4SxHBb7TFL1A:-MraX1Dk4SxHBb7TFL1B:b-ez2nmz) ([raw file](https://github.com/kube-hpc/hkube/blob/97d7c31c99320cb246efaddb121a9a5bcbcf9e82/core/pipeline-driver-queue/lib/jobs/producer.js#L67)):*
> ```JavaScript
>                 if (required > 0) {
> 
>                 }
> ```

empty if



_Code quote:_
```JavaScript
                if (required > 0) {

                }
```

**Review by NassiHarel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1469)* status: 0 of 82 files reviewed, 3 unresolved discussions (waiting on @golanha, @NassiHarel, and @yehiyam)

___
*[core/pipeline-driver-queue/lib/jobs/producer.js, line 67 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1469#-MraX1Dk4SxHBb7TFL1A:-Mra_-y270dVaPwkDu5f:b-9ej1sc) ([raw file](https://github.com/kube-hpc/hkube/blob/97d7c31c99320cb246efaddb121a9a5bcbcf9e82/core/pipeline-driver-queue/lib/jobs/producer.js#L67)):*
<details><summary><i>Previously, golanha (Golan Hallel) wrote…</i></summary><blockquote>

empty if
</blockquote></details>

I just pushed it and go back working on 1.3
___
*[core/pipeline-driver-queue/tests/test.js, line 189 at r9](https://reviewable.io/reviews/kube-hpc/hkube/1469#-MqDWVZHiInNI6Iau5Dl-r9-189:-Mra_KiQE353oyrz58Ms:b32w12) ([raw file](https://github.com/kube-hpc/hkube/blob/c36a8d55db195dcb6d3de2ab82ba11bf24c23027/core/pipeline-driver-queue/tests/test.js#L189)):*
<details><summary><i>Previously, golanha (Golan Hallel) wrote…</i></summary><blockquote>

Does the Datastore allow creating different records with the same jobId?
</blockquote></details>

No

**Review by NassiHarel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1469)* status: 0 of 132 files reviewed, 8 unresolved discussions (waiting on @golanha and @yehiyam)

___
*[core/algorithm-queue/tests/test.js, line 140 at r9](https://reviewable.io/reviews/kube-hpc/hkube/1469#-MrS3Jxv_jRzqyAMFsW4-r9-140:-Mt8aciQ2ss6gxOkQa6p:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/c36a8d55db195dcb6d3de2ab82ba11bf24c23027/core/algorithm-queue/tests/test.js#L140)):*
<details><summary><i>Previously, golanha (Golan Hallel) wrote…</i></summary><blockquote>

test left in skip
</blockquote></details>

Done.
___
*[core/pipeline-driver-queue/tests/setup.js, line 22 at r28](https://reviewable.io/reviews/kube-hpc/hkube/1469#-MuR2pvY4mS6as2fw7Iw:-MuR2pvY4mS6as2fw7Ix:bqqpi1p) ([raw file](https://github.com/kube-hpc/hkube/blob/b4b4577855365bfc537b85700335d17230e051a4/core/pipeline-driver-queue/tests/setup.js#L22)):*
> ```JavaScript
> });
> beforeEach(async () => {
>     await dataStore._db.jobs.deleteMany({}, { allowNotFound: true });
> ```

this is already imported, `stateManager`



_Code quote:_
```JavaScript
dataStore
```
___
*[core/pipeline-driver-queue/tests/test-preferred-managed.js, line 19 at r28](https://reviewable.io/reviews/kube-hpc/hkube/1469#-MuR29mlAXb5-AFQNPk5:-MuR29mlAXb5-AFQNPk6:b-c4eiie) ([raw file](https://github.com/kube-hpc/hkube/blob/b4b4577855365bfc537b85700335d17230e051a4/core/pipeline-driver-queue/tests/test-preferred-managed.js#L19)):*
> ```JavaScript
>         producerLib = require('../lib/jobs/producer');
>     });
>     beforeEach(async () => {
> ```

this is not async
also, move it to setup.js



_Code quote:_
```JavaScript
beforeEach(async () => {
```
___
*[core/pipeline-driver-queue/tests/test-preferred-managed.js, line 33 at r28](https://reviewable.io/reviews/kube-hpc/hkube/1469#-MuR26LECAkcD5dhWU37:-MuR26LECAkcD5dhWU38:b-s76nm8) ([raw file](https://github.com/kube-hpc/hkube/blob/b4b4577855365bfc537b85700335d17230e051a4/core/pipeline-driver-queue/tests/test-preferred-managed.js#L33)):*
> ```JavaScript
>         beforeEach(async () => {
>             jobs = [];
>             jobs.push({ jobId: 'a', pipeline: { name: 'p_a' }, entranceTime: 1, priority: 1 });
> ```

This is not true, it should be `pipelineName`



_Code quote:_
```JavaScript
pipeline: { name: 'p_a' }
```
___
*[core/pipeline-driver-queue/tests/test-preferred-managed.js, line 103 at r28](https://reviewable.io/reviews/kube-hpc/hkube/1469#-MuR2N5XBRlmiE4cH6vs:-MuR2N5XBRlmiE4cH6vt:bo3ql6k) ([raw file](https://github.com/kube-hpc/hkube/blob/b4b4577855365bfc537b85700335d17230e051a4/core/pipeline-driver-queue/tests/test-preferred-managed.js#L103)):*
> ```JavaScript
>         it('preferred order', async () => {
>             const jobs = [];
>             jobs.push({ jobId: 'a', pipeline: { name: 'p_a' }, entranceTime: 10, calculated: { latestScores: [] } });
> ```

you don't need this, I already added it to the `stubTemplate` function



_Code quote:_
```JavaScript
calculated: { latestScores: [] 
```
___
*[core/pipeline-driver-queue/tests/stub/stub.js, line 6 at r28](https://reviewable.io/reviews/kube-hpc/hkube/1469#-MuR2a2M8hK_fqN-Cb43:-MuR2a2M8hK_fqN-Cb44:b-2mshj) ([raw file](https://github.com/kube-hpc/hkube/blob/b4b4577855365bfc537b85700335d17230e051a4/core/pipeline-driver-queue/tests/stub/stub.js#L6)):*
> ```JavaScript
>     score: score || Math.floor((Math.random() * 100)),
>     preference,
>     pipelineName: pipeline?.name || `pipeline-${uuidv4()}`,
> ```

pass `pipelineName`
pass `tags`



_Code quote:_
```JavaScript
pipeline?.name
```

**Review by NassiHarel — COMMENTED** [review]
:-)

**Pull request closed without merging**